src/System.Net.Security/src/System/Net/Windows/SSPISecureChannelType.cs

```diff
@@ -0,0 +1,790 @@
+                if ((secStatus == SecurityStatus.OK && decspc[i].type == SecurityBufferType.Data)
+                    // or we failed to decode the data, here is the encoded data.
+                    || (secStatus != SecurityStatus.OK && decspc[i].type == SecurityBufferType.Extra))
+                {
+                    offset = decspc[i].offset;
+                    count = decspc[i].size;
+                    break;
+                }
+            }
+
+            return secStatus;
+        }
+
+        public unsafe int QueryContextChannelBinding(SafeDeleteContext phContext, ChannelBindingKind attribute, out SafeFreeContextBufferChannelBinding refHandle)
+        {
+            refHandle = SafeFreeContextBufferChannelBinding.CreateEmptyHandle();
+
+            // Bindings is on the stack, so there's no need for a fixed block.
+            Bindings bindings = new Bindings();
+            int errorCode = SafeFreeContextBufferChannelBinding.QueryContextChannelBinding(phContext, (Interop.Secur32.ContextAttribute)attribute, &bindings, refHandle);
+
+            if (errorCode != 0)
+            {
+                GlobalLog.Leave("QueryContextChannelBinding", "ERROR = " + ErrorDescription(errorCode));
+                refHandle = null;
+            }           
+
+            return errorCode;
+        }
+
+        public int QueryContextStreamSizes(SafeDeleteContext securityContext, out StreamSizes streamSizes)
+        {
+            int errorCode;
+            streamSizes = QueryContextAttributes(securityContext, Interop.Secur32.ContextAttribute.StreamSizes, out errorCode) as StreamSizes;
+            return errorCode;
+        }
+
+        public int QueryContextConnectionInfo(SafeDeleteContext securityContext, out SslConnectionInfo connectionInfo)
+        {
+            int errorCode;
+            connectionInfo = QueryContextAttributes(securityContext, Interop.Secur32.ContextAttribute.ConnectionInfo, out errorCode) as SslConnectionInfo;
+            return errorCode;
+        }
+
+        public int QueryContextRemoteCertificate(SafeDeleteContext securityContext, out SafeFreeCertContext remoteCert)
+        {
+            int errorCode;
+            remoteCert = QueryContextAttributes(securityContext, Interop.Secur32.ContextAttribute.RemoteCertificate, out errorCode) as SafeFreeCertContext;
+            return errorCode;
+        }
+
+        public int QueryContextIssuerList(SafeDeleteContext securityContext, out Object issuerList)
+        {
+            int errorCode;
+            issuerList = QueryContextAttributes(securityContext, Interop.Secur32.ContextAttribute.IssuerListInfoEx, out errorCode);
+            return errorCode;
+        }    
+
+        #region Private Methods
+
+        private enum OP
+        {
+            Encrypt = 1,
+            Decrypt
+        }
+
+        private SecurityStatus MapToSecurityStatus(Interop.SecurityStatus secStatus)
```

Comment:
This needs to either get better or removed.

We can't rely on converting an enum to string then back to an int... Perf will be horrible.